File: src/main/java/com/pfe/saas/service/ChannelService.java
package com.pfe.saas.service;

import com.pfe.saas.dto.request.ChannelMessageRequest;
import com.pfe.saas.dto.request.ChannelRequest;
import com.pfe.saas.entity.*;
import com.pfe.saas.enums.ChannelMemberRole;
import com.pfe.saas.enums.ReactionType;
import com.pfe.saas.repository.*;
import lombok.RequiredArgsConstructor;
import org.hibernate.Hibernate;
import org.springframework.data.domain.*;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ChannelService {

    private final ChannelRepository channelRepository;
    private final ChannelMemberRepository memberRepository;
    private final ChannelMessageRepository messageRepository;
    private final ChannelMessageReactionRepository reactionRepository;
    private final UserRepository userRepository;

    private User currentUser() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Utilisateur non trouvé"));
    }

    // ── Channels ──────────────────────────────────────────────────────────────

    @Transactional
    public Channel createChannel(ChannelRequest req) {
        User owner = currentUser();
        Channel channel = new Channel();
        channel.setName(req.getName());
        channel.setDescription(req.getDescription());
        channel.setPictureUrl(req.getPictureUrl());
        channel.setPublicChannel(req.isPublicChannel());
        channel.setOwner(owner);
        Channel saved = channelRepository.save(channel);

        // Owner is automatically a member
        ChannelMember ownerMember = new ChannelMember();
        ownerMember.setChannel(saved);
        ownerMember.setUser(owner);
        ownerMember.setRole(ChannelMemberRole.OWNER);
        memberRepository.save(ownerMember);

        return saved;
    }

    @Transactional
    public Channel updateChannel(Long channelId, ChannelRequest req) {
        Channel channel = getChannelOrThrow(channelId);
        requireOwnerOrAdmin(channel);
        channel.setName(req.getName());
        channel.setDescription(req.getDescription());
        channel.setPictureUrl(req.getPictureUrl());
        channel.setPublicChannel(req.isPublicChannel());
        return channelRepository.save(channel);
    }

    @Transactional
    public void deleteChannel(Long channelId) {
        Channel channel = getChannelOrThrow(channelId);
        requireOwner(channel);
        channelRepository.delete(channel);
    }

    public Page<Channel> searchPublicChannels(String query, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        if (query == null || query.isBlank()) {
            return channelRepository.findByPublicChannelTrue(pageable);
        }
        return channelRepository.searchPublic(query, pageable);
    }

    public Page<Channel> getMyChannels(int page, int size) {
        User me = currentUser();
        return channelRepository.findJoinedByUserId(me.getId(), PageRequest.of(page, size));
    }

    public Channel getChannelById(Long channelId) {
        Channel channel = getChannelOrThrow(channelId);
        channel.setMemberCount(memberRepository.countByChannelId(channelId));
        User me = currentUser();
        channel.setMember(memberRepository.existsByChannelIdAndUserId(channelId, me.getId()));
        return channel;
    }

    // ── Membership ────────────────────────────────────────────────────────────

    @Transactional
    public void join(Long channelId) {
        Channel channel = getChannelOrThrow(channelId);
        if (!channel.isPublicChannel()) throw new RuntimeException("Canal privé");
        User me = currentUser();
        if (memberRepository.existsByChannelIdAndUserId(channelId, me.getId())) return;
        ChannelMember member = new ChannelMember();
        member.setChannel(channel);
        member.setUser(me);
        member.setRole(ChannelMemberRole.MEMBER);
        memberRepository.save(member);
    }

    @Transactional
    public void leave(Long channelId) {
        Channel channel = getChannelOrThrow(channelId);
        User me = currentUser();
        if (channel.getOwner().getId().equals(me.getId())) {
            throw new RuntimeException("Le propriétaire ne peut pas quitter son canal");
        }
        memberRepository.deleteByChannelIdAndUserId(channelId, me.getId());
    }

    public List<ChannelMember> getMembers(Long channelId) {
        assertMember(channelId);
        return memberRepository.findByChannelId(channelId);
    }

    // ── Messages ──────────────────────────────────────────────────────────────

    @Transactional
    public ChannelMessage sendMessage(Long channelId, ChannelMessageRequest req) {
        Channel channel = getChannelOrThrow(channelId);
        User me = currentUser();
        if (!memberRepository.existsByChannelIdAndUserId(channelId, me.getId())) {
            throw new RuntimeException("Vous n'êtes pas membre de ce canal");
        }
        ChannelMessage msg = new ChannelMessage();
        msg.setChannel(channel);
        msg.setSender(me);
        msg.setContent(req.getContent());
        msg.setAttachmentUrl(req.getAttachmentUrl());
        if (req.getReplyToId() != null) {
            ChannelMessage replyTo = messageRepository.findById(req.getReplyToId()).orElse(null);
            msg.setReplyTo(replyTo);
        }
        return messageRepository.save(msg);
    }

    @Transactional(readOnly = true)
    public Page<ChannelMessage> getMessages(Long channelId, int page, int size) {
        assertMember(channelId);
        Page<ChannelMessage> messages = messageRepository.findByChannelIdOrderBySentAtAsc(
                channelId, PageRequest.of(page, size));
        messages.getContent().forEach(msg -> {
            if (msg.getReplyTo() != null) {
                Hibernate.initialize(msg.getReplyTo());
            }
            msg.setReactionCount(reactionRepository.countByChannelMessageId(msg.getId()));
        });
        return messages;
    }

    @Transactional
    public void deleteMessage(Long messageId) {
        ChannelMessage msg = messageRepository.findById(messageId)
                .orElseThrow(() -> new RuntimeException("Message non trouvé"));
        User me = currentUser();
        if (!msg.getSender().getId().equals(me.getId())) {
            throw new RuntimeException("Accès non autorisé");
        }
        messageRepository.delete(msg);
    }

    // ── Réactions ─────────────────────────────────────────────────────────────

    @Transactional
    public void reactToMessage(Long messageId, ReactionType type) {
        ChannelMessage msg = messageRepository.findById(messageId)
                .orElseThrow(() -> new RuntimeException("Message non trouvé"));
        User me = currentUser();
        reactionRepository.findByChannelMessageIdAndUserId(messageId, me.getId())
                .ifPresentOrElse(
                        r -> { r.setType(type); reactionRepository.save(r); },
                        () -> {
                            ChannelMessageReaction r = new ChannelMessageReaction();
                            r.setChannelMessage(msg);
                            r.setUser(me);
                            r.setType(type);
                            reactionRepository.save(r);
                        }
                );
    }

    @Transactional
    public void removeReaction(Long messageId) {
        User me = currentUser();
        reactionRepository.deleteByChannelMessageIdAndUserId(messageId, me.getId());
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private Channel getChannelOrThrow(Long channelId) {
        return channelRepository.findById(channelId)
                .orElseThrow(() -> new RuntimeException("Canal non trouvé"));
    }

    private void requireOwner(Channel channel) {
        User me = currentUser();
        if (!channel.getOwner().getId().equals(me.getId())) {
            throw new RuntimeException("Seul le propriétaire peut effectuer cette action");
        }
    }

    private void requireOwnerOrAdmin(Channel channel) {
        User me = currentUser();
        if (channel.getOwner().getId().equals(me.getId())) return;
        memberRepository.findByChannelIdAndUserId(channel.getId(), me.getId())
                .filter(m -> m.getRole() == ChannelMemberRole.OWNER || m.getRole() == ChannelMemberRole.ADMIN)
                .orElseThrow(() -> new RuntimeException("Droits insuffisants"));
    }

    private void assertMember(Long channelId) {
        User me = currentUser();
        if (!memberRepository.existsByChannelIdAndUserId(channelId, me.getId())) {
            throw new RuntimeException("Vous n'êtes pas membre de ce canal");
        }
    }
}

File: src/app/core/services/channel.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse, Channel, ChannelMember, ChannelMessage, ChannelMessageRequest, ChannelRequest, Page } from '../models/models';

@Injectable({ providedIn: 'root' })
export class ChannelService {
  private readonly http = inject(HttpClient);
  private readonly API = `${environment.apiUrl}/channels`;

  // ── Channels ──────────────────────────────────────────────────────────────
  create(req: ChannelRequest): Observable<ApiResponse<Channel>> {
    return this.http.post<ApiResponse<Channel>>(this.API, req);
  }

  update(id: number, req: ChannelRequest): Observable<ApiResponse<Channel>> {
    return this.http.put<ApiResponse<Channel>>(`${this.API}/${id}`, req);
  }

  delete(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.API}/${id}`);
  }

  search(q?: string, page = 0, size = 20): Observable<ApiResponse<Page<Channel>>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (q) params = params.set('q', q);
    return this.http.get<ApiResponse<Page<Channel>>>(this.API, { params });
  }

  getMyChannels(page = 0, size = 20): Observable<ApiResponse<Page<Channel>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<Page<Channel>>>(`${this.API}/mine`, { params });
  }

  getById(id: number): Observable<ApiResponse<Channel>> {
    return this.http.get<ApiResponse<Channel>>(`${this.API}/${id}`);
  }

  // ── Membership ────────────────────────────────────────────────────────────
  join(id: number): Observable<ApiResponse<void>> {
    return this.http.post<ApiResponse<void>>(`${this.API}/${id}/join`, {});
  }

  leave(id: number): Observable<ApiResponse<void>> {
    return this.http.post<ApiResponse<void>>(`${this.API}/${id}/leave`, {});
  }

  getMembers(id: number): Observable<ApiResponse<ChannelMember[]>> {
    return this.http.get<ApiResponse<ChannelMember[]>>(`${this.API}/${id}/members`);
  }

  // ── Messages ──────────────────────────────────────────────────────────────
  sendMessage(id: number, req: ChannelMessageRequest): Observable<ApiResponse<ChannelMessage>> {
    return this.http.post<ApiResponse<ChannelMessage>>(`${this.API}/${id}/messages`, req);
  }

  getMessages(id: number, page = 0, size = 50): Observable<ApiResponse<Page<ChannelMessage>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<Page<ChannelMessage>>>(`${this.API}/${id}/messages`, { params });
  }

  deleteMessage(messageId: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.API}/messages/${messageId}`);
  }

  // ── Réactions ─────────────────────────────────────────────────────────────
  react(messageId: number, type: string): Observable<ApiResponse<void>> {
    return this.http.post<ApiResponse<void>>(`${this.API}/messages/${messageId}/react`, { type });
  }

  removeReaction(messageId: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.API}/messages/${messageId}/react`);
  }
}
